homeassistant/components/alpha_vantage/__init__.py

```diff
@@ -1 +1,167 @@
-"""The alpha_vantage component."""
+"""Support for stock market information from Alpha Vantage."""
+import logging
+from datetime import timedelta
+import re
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_API_KEY, CONF_CURRENCY, CONF_NAME, \
+    CONF_SCAN_INTERVAL
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.discovery import load_platform
+from homeassistant.helpers.dispatcher import dispatcher_send
+from homeassistant.helpers.event import track_time_interval
+from homeassistant.components.sensor import DOMAIN as SENSOR
+
+REQUIREMENTS = ['alpha_vantage==2.1.0']
+
+DOMAIN = 'alpha_vantage'
+STOCK_DATA_UPDATED = '{}_stock_data_updated'.format(DOMAIN)
+FOREX_DATA_UPDATED = '{}_forex_data_updated'.format(DOMAIN)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_FOREIGN_EXCHANGE = 'foreign_exchange'
+CONF_FROM = 'from'
+CONF_SYMBOL = 'symbol'
+CONF_SYMBOLS = 'symbols'
+CONF_TO = 'to'
+
+DEFAULT_SCAN_INTERVAL = timedelta(minutes=5)
+
+SYMBOL_SCHEMA = vol.Schema({
+    vol.Required(CONF_SYMBOL): cv.string,
+    vol.Optional(CONF_CURRENCY): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+CURRENCY_SCHEMA = vol.Schema({
+    vol.Required(CONF_FROM): cv.string,
+    vol.Required(CONF_TO): cv.string,
+    vol.Optional(CONF_NAME): cv.string,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(
+        vol.Schema({
+            vol.Required(CONF_API_KEY): cv.string,
+            vol.Optional(CONF_FOREIGN_EXCHANGE, default=[]):
+                vol.All(cv.ensure_list, [CURRENCY_SCHEMA]),
+            vol.Optional(CONF_SYMBOLS, default=[]):
+                vol.All(cv.ensure_list, [SYMBOL_SCHEMA]),
+            vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL):
+                vol.All(cv.time_period, cv.positive_timedelta),
+        }),
+        cv.has_at_least_one_key(CONF_FOREIGN_EXCHANGE, CONF_SYMBOLS)
+    )
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Alpha Vantage component."""
+    conf = config[DOMAIN]
+    data = hass.data[DOMAIN] = AlphaVantageData(
+        hass,
+        conf[CONF_API_KEY],
+        [symbol[CONF_SYMBOL] for symbol in conf[CONF_SYMBOLS]],
+        conf[CONF_FOREIGN_EXCHANGE]
+    )
+    data.update()
```

Comment:
What happens if the api key is invalid? Can we catch that and fail set up?

Reply:
The issue here is that we're very limited in API calls - 5/minute, 500/day and it does cost us one API call to validate the key.

Reply:
But it will only happen during set up, so after a restart of home assistant. That should be ok, right?